### PR TITLE
chore: update PTL adapter to support latest PTL [DET-6421]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

Our current PTL adapter doesn't work out-of-the-box with the latest PTL. We needed to upgrade PTL because of a security fix. 

Main changes are in `_adapter.py` and PTL examples (`mnist_pl` and `gan_mnist_pl`). Default images were changed to test the new image builds.

What changed:
- `TrainerOptimizersMixin` now requires a Trainer() module instance, so the same logic of `init_optimizers` is replicated in our adapter to circumnavigate the requirement.
- `torchmetrics` is where all the accuracy functions live in PTL now
- There's a bug in `save_hyperparameters` that leaks hyperparameters. Workaround is to explicitly state which hyperparameters to save.

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by ioga** [comment]
Does the new code still work with the old pytorch-lightning==1.3.5?
If not, could we please add a release note announcing this forced upgrade / breaking change, and maybe mention the supported pytorch lightning version in the docs somewhere?

File: harness/determined/pytorch/lightning/_adapter.py [hunk 1]
```diff
@@ -315,6 +317,23 @@ def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
...
+        Configure and validate optimizers and schedulers in pytorch lightning.
+        This implements the functionality of
+        pytorch_lightning.trainer.optimizers.TrainerOptimizersMixin but circumvents the Trainer
+        initialization requirement.
+        """
+        optimizer_conf = self._pls.lm.configure_optimizers()
+        optimizers, lr_schedulers, _, monitor = TrainerOptimizersMixin._configure_optimizers(
+            optimizer_conf
+        )
+        lr_schedulers = TrainerOptimizersMixin._configure_schedulers(
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/lightning/_adapter.py:331]: ditto on this method

File: harness/determined/pytorch/lightning/_adapter.py [hunk 2]
```diff
@@ -315,6 +317,23 @@ def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
...
+            lr_schedulers, monitor, not self._pls.lm.automatic_optimization
+        )
+        _validate_scheduler_optimizer(optimizers, lr_schedulers)
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/lightning/_adapter.py:334]: and ditto on this one too

File: examples/gan/gan_mnist_pl/gan.py [hunk 3]
```diff
@@ -113,7 +113,8 @@ def __init__(
     def forward(self, z):
         return self.generator(z)
 
-    def adversarial_loss(self, y_hat, y):
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (azhou-determined) [reply examples/gan/gan_mnist_pl/gan.py:116]: it's because the PTL adversarial_loss method API changed, and if we don't conform to it, I get an override check error about the parent not defining the method.

File: harness/determined/pytorch/lightning/_adapter.py [hunk 4]
```diff
@@ -315,6 +317,23 @@ def lightning_scheduler_dict_to_det(lrs: dict) -> _LRScheduler:
...
 
+    def _init_optimizers(self) -> Tuple[List, List]:
+        """
+        Configure and validate optimizers and schedulers in pytorch lightning.
+        This implements the functionality of
+        pytorch_lightning.trainer.optimizers.TrainerOptimizersMixin but circumvents the Trainer
+        initialization requirement.
+        """
+        optimizer_conf = self._pls.lm.configure_optimizers()
+        optimizers, lr_schedulers, _, monitor = TrainerOptimizersMixin._configure_optimizers(
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/pytorch/lightning/_adapter.py:328]: Ok, I see it there in v1.5.9, but I don't see it in tip-of-master for pytorch lightning.  Am I crazy?

Otherwise this is a big red-flag, that we're depending on internals that are constantly changing

[the remark this answers is not in the document]
Comment (azhou-determined) [reply harness/determined/pytorch/lightning/_adapter.py:328]: traced the commit history in PTL, looks like they just moved this logic to pytorch_lightning.core from pytorch_lightning.trainer (https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/optimizer.py). 
so i think this change is pretty safe, seeing as their internal logic hasn't changed. future upgrades just need to migrate where the static `_ configure_optimizers` call comes from in PTL.

**Review by ioga — APPROVED** [review]
lgtm - I've tried this PR out and it worked for me.

before the merge, we still should:
- switch bumpenvs to the proper tags
- add a release note

File: harness/determined/pytorch/lightning/_adapter.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/pytorch/lightning/_adapter.py:328]: That's not "safe", it means that it's going to be a lot of work to continue to work against multiple versions of PTL.

I suppose we can deal with that later.  I don't see a point in rearchitecting our PTL support today.

**Review by rb-determined-ai — COMMENTED** [review]
You have my "soft approval", but we definitely need to work out the image tags you've modified in this pr before landing
